sql/src/main/java/io/crate/operation/merge/PassThroughPagingIterator.java

```diff
@@ -37,15 +71,26 @@
     }
 
     @Override
-    public void merge(Iterable<? extends Iterator<T>> iterators) {
+    public void merge(Iterable<? extends Iterable<T>> iterables) {
+        if (repeatable) {
+            this.iterables.addAll(iterables);
+        }
         if (iterator.hasNext()) {
-            iterator = Iterators.concat(iterator, Iterators.concat(iterators.iterator()));
+            iterator = Iterators.concat(iterator,
+                    Iterators.concat(Iterables.transform(iterables, TO_ITERATOR).iterator()));
         } else {
-            iterator = Iterators.concat(iterators.iterator());
+            iterator = Iterators.concat(Iterables.transform(iterables, TO_ITERATOR).iterator());
         }
     }
 
     @Override
     public void finish() {
     }
+
+    @Override
+    public Iterator<T> repeat() {
+        return Iterators.concat(
+                Iterables.transform(this.iterables.build(), TO_ITERATOR).iterator()
```

Comment:
maybe you could do the concat on the Iterables and save the whole result (concat + build) in a local variable and use that for additional repeats